### PR TITLE
DM-15098: Add Registry.getRegion(DataId)

### PR DESCRIPTION
**Pull request by pschella**

### CODE REVIEW

File: python/lsst/daf/butler/registries/sqlRegistry.py [hunk 1]
```diff
@@ -992,6 +994,42 @@ def find(self, collection, datasetType, dataId):
...
+            within a Collection.
+
+        Returns
+        -------
+        region : `lsst.sphgeom.ConvexPolygon`
+            The region associated with a `dataId`.
+
+        Raises
+        ------
+        `ValueError`
```
Comment (timj) [inline python/lsst/daf/butler/registries/sqlRegistry.py]: Exception names in a Raises block are not backtick quoted.

File: tests/test_sqlRegistry.py [hunk 2]
```diff
@@ -376,6 +377,17 @@ def testBasicTransaction(self):
         # Or the DataUnit entries
         self.assertIsNone(registry.findDataUnitEntry("Camera", {"camera": "DummyCam"}))
 
+    def testGetRegion(self):
+        registry = Registry.fromConfig(self.butlerConfig, create=True)
+        region = lsst.sphgeom.ConvexPolygon((lsst.sphgeom.UnitVector3d(1, 0, 0),
+                                             lsst.sphgeom.UnitVector3d(0, 1, 0),
+                                             lsst.sphgeom.UnitVector3d(0, 0, 1)))
+        registry.addDataUnitEntry("SkyMap", {"skymap": "DummySkyMap", "sha1": bytes()})
```
Comment (timj) [inline tests/test_sqlRegistry.py]: I've just noticed we are insisting on this being a `sha1`. Would it generically be better to call this `hash` or something?

File: tests/test_sqlRegistry.py [hunk 3]
```diff
@@ -376,6 +377,17 @@ def testBasicTransaction(self):
...
+        registry.addDataUnitEntry("Tract", {"skymap": "DummySkyMap", "tract": 0, "region": region})
+        dataId = {"skymap": "DummySkyMap", "tract": 0}
```
Comment (timj) [inline tests/test_sqlRegistry.py]: Can you also show that `tract: 1` doesn't get retrieved?

File: python/lsst/daf/butler/registries/sqlRegistry.py [hunk 4]
```diff
@@ -992,6 +994,42 @@ def find(self, collection, datasetType, dataId):
         else:
             return None
 
+    def getRegion(self, dataId):
+        """Get region associated with a dataId.
+
+        Parameters
+        ----------
+        dataId : `dict`
+            A `dict` of `DataUnit` name, value pairs that label the `DatasetRef`
```
Comment (timj) [inline python/lsst/daf/butler/registries/sqlRegistry.py]: Is it confusing to call out to the `DataUnit` class docs here even though these arguments are not of that class?

File: python/lsst/daf/butler/registries/sqlRegistry.py [hunk 5]
```diff
@@ -992,6 +994,42 @@ def find(self, collection, datasetType, dataId):
...
+            If there is no spatial `DataUnit` in the provided ``dataId``.
+        `NotImplementedError`
+            If more than one `DataUnit` in the ``dataId`` is spatial.
+        """
+        haveSpatialDataUnit = False
+        for key in dataId:
+            dataUnit = self._schema.dataUnits.getByLinkName(key)
+            if dataUnit.regionColumn is not None:
+                if haveSpatialDataUnit:
+                    raise NotImplementedError("Cannot have more than one spatial region")
```
Comment (timj) [inline python/lsst/daf/butler/registries/sqlRegistry.py]: Please add a test for this condition.

**Review by timj — APPROVED** [review]
Minor comments regarding docstrings and test coverage but otherwise fine.

File: python/lsst/daf/butler/core/dataUnit.py [hunk 6]
```diff
@@ -514,3 +517,18 @@ def getPrimaryKeyNames(self, dataUnitNames):
...
+        Parameters
+        ----------
+        name : `str`
+            Link name.
+
+        Returns
+        -------
+        dataUnit : `DataUnit`
+            The corresponding `DataUnit` instance.
+        """
```
Comment (andy-slac) [inline python/lsst/daf/butler/core/dataUnit.py]: Add `Raises` section?

File: tests/test_sqlRegistry.py (hunk 2 above)
Comment (pschella) [reply tests/test_sqlRegistry.py]: Yes, but out of scope for this ticket.

File: python/lsst/daf/butler/registries/sqlRegistry.py (hunk 4 above)
Comment (pschella) [reply python/lsst/daf/butler/registries/sqlRegistry.py]: It is, and that should be fixed in multiple places. I'll do it on this ticket.

**Pull request merged**